### PR TITLE
Improved RISC-V panic dump

### PR DESCRIPTION
**Pull request by phil-levis**
### Pull Request Overview

This pull request updates the RISC-V panic handler to print some machine state (in a manner similar to the Cortex-M).

The major limitation it points out is that, due to current software structure, the kernel will print out detailed machine state if and only if there are processes. It depends on calling `fault_fmt` on `Process` in order to get access to `Chip::userspace_kernel_boundary()` to actually get/print the state.

To get around this, this PR hacks up a static mut reference `CHIP` to the chip. So this PR should probably become "fixes panic dump dependencies", in which RISC-V support is just a detail.

### Testing Strategy

This pull request was tested by compiling and running on OpenTitan. It will not compile for other boards because it changes some core interfaces.

### TODO or Help Wanted

This pull request still needs a new architecture for printing machine state when there are no userspace processes.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make formatall`.

### CODE REVIEW

File: boards/opentitan/src/alarm_test.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+#![allow(unused_imports,dead_code)]
+
+use capsules::test::alarm::TestAlarm;
+use kernel::hil::time::Alarm;
+use ibex::timer;
+use kernel::static_init;
+
+pub unsafe fn run_alarm() {
```
Comment (alistair23) [inline boards/opentitan/src/alarm_test.rs]: I don't see any mention of this in the PR description, why add an alarm test?

File: arch/rv32i/src/syscall.rs [hunk 2]
```diff
@@ -415,13 +416,166 @@ impl kernel::syscall::UserspaceKernelBoundary for SysCall {
...
+            csr::CSR.mie.get(),     csr::CSR.mip.get(),
+            e_usoft, p_usoft,
+            e_ssoft, p_ssoft,
+            e_msoft, p_msoft,
+            e_utimer, p_utimer,
+            e_stimer, p_stimer,
+            e_mtimer, p_mtimer,
+            e_uext, p_uext,
+            e_sext, p_sext,
+            e_mext, p_mext));
```
Comment (alistair23) [inline arch/rv32i/src/syscall.rs]: This looks really useful.

File: chips/ibex/src/chip.rs [hunk 3]
```diff
@@ -109,50 +109,27 @@ impl kernel::Chip for Ibex {
...
+        rv32i::csr::mcause::Exception::LoadMisaligned |
+        rv32i::csr::mcause::Exception::LoadFault |
+        rv32i::csr::mcause::Exception::StoreMisaligned |
+        rv32i::csr::mcause::Exception::StoreFault |
+        rv32i::csr::mcause::Exception::MachineEnvCall |
+        rv32i::csr::mcause::Exception::InstructionPageFault |
+        rv32i::csr::mcause::Exception::LoadPageFault |
+        rv32i::csr::mcause::Exception::StorePageFault |
+        rv32i::csr::mcause::Exception::Unknown => {
+            panic!("fatal exception");
```
Comment (alistair23) [inline chips/ibex/src/chip.rs]: Good fix!

**Comment by bradjc** [comment]
Adding a function to the `Chip` trait for printing fault debug information would be one way to make this more general. I think adding a chip parameter to the panic functions in debug.rs in the kernel make sense.

**Comment by pfmooney** [comment]
> Adding a function to the `Chip` trait for printing fault debug information would be one way to make this more general. I think adding a chip parameter to the panic functions in debug.rs in the kernel make sense.

The addition to the `Chip` trait sounds like a good idea.  Getting a reference to the chip in panic context seems a little bit trickier.  Would you want that trait method to emit a function pointer which the debug infrastructure could automatically call?

File: boards/opentitan/src/alarm_test.rs (hunk 1 above)
Comment (phil-levis) [reply boards/opentitan/src/alarm_test.rs]: It will likely not be in the in the final PR, but it's here so it's easy to test the alarm.

**Comment by phil-levis** [comment]
>  Getting a reference to the chip in panic context seems a little bit trickier. Would you want that trait method to emit a function pointer which the debug infrastructure could automatically call?

Yes, exactly. The question is how to get a reference. We could just make a global function, but this would call for a new Capability. I'm not sure the Chip trait is the right granularity; I'd like things to be able to print machine state without necessarily allowing them to call everything in `UserspaceKernelBoundary`.

**Comment by alistair23** [comment]
There could be a new `MachineState` trait that has the debug information (and whatever else is added in the future). It can be static and maybe even created by the `Chip` trait?

**Comment by bradjc** [comment]
Boards and the kernel scheduler have references, and then the chip itself would have access, what else would want a reference? Rust panics are handled by boards, so if the kernel crashes the board has a reference to the chip.

The access control is an interesting question. If we had something like `MachineState` trait, then we could mimic what we do with the MPU and allow the chip to give us a reference to just the `MachineState` object.

**Comment by phil-levis** [comment]
> Rust panics are handled by boards, so if the kernel crashes the board has a reference to the chip.

Sort of. This is essentially what the current approach does. The board defines a static mut Option for Chip, which`panic_fmt()` unwraps. What I'm hoping for (and which the `MachineState` proposal is starting to get at) is a more principled design/decomposition for these traits.

**Comment by phil-levis** [comment]
I'm closing this for a PR focused on refactoring the traits for separate process dumps and machine state dumps ( #1590 )

**Pull request closed without merging**